prusti-viper/src/encoder/spec_encoder.rs

```diff
@@ -171,6 +172,111 @@ impl<'p, 'v: 'p, 'tcx: 'v> SpecEncoder<'p, 'v, 'tcx> {
                     .collect(),
                 self.encode_assertion(body),
             ),
+            box typed::AssertionKind::SpecEnt(ref cl, ref vars, ref pres, ref posts) => {
+                let tcx = self.encoder.env().tcx();
+                let (mir, _) = tcx.mir_promoted(ty::WithOptConstParam::unknown(cl.expr));
+                let result = &mir.borrow().local_decls[(0 as u32).into()];
+                let ty = result.ty;
+                let typaram_repl = self.encoder.typaram_repl.borrow();
+                if let Some(Some(ty_repl)) = typaram_repl.iter().last().map(|x| x.get(ty)) {
+                    debug!("spec ent repl: {:?} -> {:?}", ty, ty_repl);
+
+                    match ty_repl.kind() {
+                        ty::TyKind::Closure(def_id, substs) => {
+                            let encoded_pres = pres.iter()
+                                .map(|x| self.encode_assertion(x))
+                                .collect::<Vec<vir::Expr>>()
+                                .into_iter()
+                                .conjoin();
+
+                            // encode_forall_arg() above only works for integers.
+                            // Therefore, for the time being, check that we're working with integers:
+                            vars.args.iter().map(|(arg, arg_ty)| {
+                                match arg_ty.kind() {
+                                    ty::TyKind::Int(..) | ty::TyKind::Uint(..) => {}
+                                    _ => { unimplemented!("Only integers are currently supported as closure arguments."); }
+                                }
+                            });
+                            match vars.result.1.kind() {
+                                ty::TyKind::Int(..) | ty::TyKind::Uint(..) => {}
+                                _ => { unimplemented!("Only integers are currently supported as closure return types."); }
+                            }
+
+                            let sf_pre_name = self.encoder.encode_spec_func_name(*def_id, SpecFunctionKind::Pre);
+                            let pre_conjunct = vir::Expr::forall(
+                                vars.args.iter().map(|(arg, arg_ty)| self.encode_forall_arg(*arg, arg_ty, &format!("{}_{}", vars.spec_id, vars.pre_id))).collect(),
+                                vec![], // TODO: encode triggers
+                                vir::Expr::implies(
+                                    encoded_pres.clone(),
+                                    vir::Expr::FuncApp(
+                                        sf_pre_name,
+                                        (0 .. vars.args.len())
+                                            .map(|i| vir::Expr::Local(
+                                                vir::LocalVar::new(format!("_{}_forall", i + 2), vir::Type::Int),
```

Comment:
Should this call `encode_forall_arg` like 8 lines above?

Reply:
You are right, the forall encoding has change since this was written. This fixes part of the issue.

Reply:
Indeed, that was part of the problem. Thanks for the hint @fpoli!